lib/cn/spill.c

```diff
@@ -180,394 +180,439 @@ kv_spill_abort(struct cn_tstate_omf *omf, void *arg)
 
-    u64 dbg_prev_seq HSE_MAYBE_UNUSED;
-    uint dbg_prev_src HSE_MAYBE_UNUSED;
-    uint dbg_nvals_this_key HSE_MAYBE_UNUSED;
-    bool dbg_dup HSE_MAYBE_UNUSED;
+        /* TODO: Remove HSE_MPOLICY_AGE_INTERNAL once we move to a strict 2-level cN tree */
+        pnode = w->cw_node;
+        if (w->cw_action == CN_ACTION_SPILL) {
+            if (is_spill_to_intnode(pnode, cnum))
+                kvset_builder_set_agegroup(bldr, HSE_MPOLICY_AGE_INTERNAL);
+            else
+                kvset_builder_set_agegroup(bldr, HSE_MPOLICY_AGE_LEAF);
+        } else {
+            if (cn_node_isleaf(pnode))
+                kvset_builder_set_agegroup(bldr, HSE_MPOLICY_AGE_LEAF);
+            if (cn_node_isroot(pnode))
+                kvset_builder_set_agegroup(bldr, HSE_MPOLICY_AGE_ROOT);
+            else
+                kvset_builder_set_agegroup(bldr, HSE_MPOLICY_AGE_INTERNAL);
+        }
+    }
 
-    if (w->cw_prog_interval && w->cw_progress)
-        tprog = jiffies;
+    *bldr_out = bldr;
 
-    err = merge_init(&bh, w->cw_inputv, w->cw_kvset_cnt, &w->cw_stats);
-    if (ev(err))
-        return err;
+    return err;
+}
 
-    more = get_next_item(bh, w->cw_inputv, &curr, &w->cw_stats, &err);
-    if (!more || ev(err))
-        goto done;
+/*
+ * [HSE_REVISIT] direct read path allocates buffer. Performing direct reads into the buffer
+ * in kvset builder without this is a future opportunity.
+ */
+static merr_t
+get_direct_read_buf(uint len, bool aligned_voff, u32 *bufsz, void **buf)
+{
+    uint bufsz_min = len;
 
-    cn_sfx_len = w->cw_cp->sfx_len;
+    if (ev(len > HSE_KVS_VALUE_LEN_MAX)) {
+        assert(len <= HSE_KVS_VALUE_LEN_MAX);
+        return merr(EBUG);
+    }
 
-    /* We must issue a direct read for all values that will not fit into
-     * the vblock readahead buffer.  Since all direct reads require page
-     * size alignment any value whose length is greater than the buffer
-     * size minus one page must be read directly from disk (vs from the
-     * readahead buffer).
+    /* If value offset is not page aligned then allocate one additional page to prevent a
+     * potential copy inside direct read function
      */
-    direct_read_len = w->cw_rp->cn_compact_vblk_ra;
-    direct_read_len -= PAGE_SIZE;
+    if (!aligned_voff)
+        bufsz_min += PAGE_SIZE;
 
-    tstart = perfc_ison(w->cw_pc, PERFC_DI_CNCOMP_VGET) ? 1 : 0;
+    if (!(*buf) || *bufsz < bufsz_min) {
+        const uint vlen_max = HSE_KVS_VALUE_LEN_MAX;
 
-new_key:
-    pt_spread = 0;
-    childmask = 0;
+        *bufsz = (len < vlen_max / 4) ? vlen_max / 4 :
+            ((len < vlen_max / 2) ? vlen_max / 2 : vlen_max);
 
-    if (atomic_read(w->cw_cancel_request)) {
-        err = merr(ESHUTDOWN);
-        goto done;
-    }
+        /* add an extra page if not aligned */
+        if (bufsz_min < *bufsz)
+            *bufsz += PAGE_SIZE;
 
-    if (tprog) {
-        u64 now = jiffies;
+        free_aligned(*buf);
```

Comment:
Seems like we should just be using C11 aligned_alloc()

Reply:
I vaguely remember that we changed alloc_aligned() to aligned_alloc() only in the non hot paths. Right? Maybe this code was left behind for a reason, not sure...

Reply:
@beckerg would know best. Surely he can tell us when he gets back! Not sure if he ever did benchmarking though.

Reply:
Yep, aligned_alloc() is pretty awful in hot paths, which is why aligned_alloc() still exists.  Not obvious how hot this path is since I can't see the code at the moment...